crawlfrontier/contrib/canonicalsolvers/basic.py

```diff
@@ -0,0 +1,44 @@
+# -*- coding: utf-8 -*-
+from crawlfrontier.core.components import CanonicalSolver
+
+
+class BasicCanonicalSolver(CanonicalSolver):
+    """
+    Implements a simple CanonicalSolver taking always first URL from redirect chain, if there were redirects.
+    It allows easily to avoid leaking of requests in CF (e.g. when request issued by
+    :attr:`get_next_requests() <crawlfrontier.core.manager.FrontierManager.get_next_requests>` never matched in
+    :attr:`page_crawled() <crawlfrontier.core.manager.FrontierManager.page_crawled>`) at the price of duplicating
+    records in CF for pages having more than one URL or complex redirects chains.
+    """
+    def frontier_start(self):
+        pass
+
+    def frontier_stop(self):
+        pass
+
+    def add_seeds(self, seeds):
+        pass
+
+    def page_crawled(self, response, links):
+        pass
+
+    def request_error(self, page, error):
+        pass
+
+    def get_canonical_url(self, obj):
+        canonical_url = obj.url
+        canonical_fingerprint = obj.meta['fingerprint']
+        canonical_domain_fingerprint = obj.meta['domain']['fingerprint'] if 'domain' in obj.meta else str()
+        if 'redirect_urls' in obj.meta:
+            redirect_urls = obj.meta['redirect_urls']
+            redirect_fingerprints = obj.meta['redirect_fingerprints']
+            redirect_domains = obj.meta['redirect_domains']
```

Comment:
Ah, I see, it's more like a list of triples `[(url, fprint, domain),]` rather than 3 lists of values, you're using it to skip extracting a domain and getting a fingerprint for `page_crawled` event (and for `request_error`). 
Ok, I'm just worrying if we really need to get all these values for each request url in a redirects chain, usually we need only one of them and it can be overkill, as `tldextract` is not very fast.
Probably it makes sense to keep only the first and the last, and a chain length, but it's more like a premature optimization for me now, I propose to follow "make it work -> make it right -> make it fast" and now we're somewhere near 2nd point :)